element_array_ephys/export/nwb/nwb.py

```diff
@@ -0,0 +1,605 @@
+    dict
+
+    """
+    return {
+        (electrodes["group"][idx].device.name, electrodes["id_in_probe"][idx],): idx
+        for idx in range(len(electrodes))
+    }
+
+
+def gains_helper(gains):
+    if all(x == 1 for x in gains):
+        return dict(conversion=1e-6, channel_conversion=None)
+    if all(x == gains[0] for x in gains):
+        return dict(conversion=1e-6 * gains[0], channel_conversion=None)
+    return dict(conversion=1e-6, channel_conversion=gains)
+
+
+def add_ephys_recording_to_nwb(
+    session_key: dict, nwbfile: pynwb.NWBFile, end_frame: int = None
+):
+    """Read voltage data directly from source files and iteratively transfer them to the NWB file. Automatically
+    applies lossless compression to the data, to the final file might be smaller than the original, but there is no
+    data loss. Currently supports neuropixel and openephys, and relies on SpikeInterface to read the data.
+
+    source data -> acquisition["ElectricalSeries"]
+
+    Parameters
+    ----------
+    session_key: dict
+    nwbfile: NWBFile
+    end_frame: int, optional
+        Used for small test conversions
+    """
+
+    if nwbfile.electrodes is None:
+        add_electrodes_to_nwb(session_key, nwbfile)
+
+    mapping = get_electrodes_mapping(nwbfile.electrodes)
+
+    for ephys_recording_record in (ephys.EphysRecording & session_key).fetch(
+        as_dict=True
+    ):
+        probe_id = (ephys.ProbeInsertion() & ephys_recording_record).fetch1("probe")
+
+        relative_path = (
+            ephys.EphysRecording.EphysFile & ephys_recording_record
+        ).fetch1("file_path")
+        relative_path = relative_path.replace("\\","/")
+        file_path = ephys.find_full_path(ephys.get_ephys_root_data_dir(), relative_path)
+
+        if ephys_recording_record["acq_software"] == "SpikeGLX":
+            extractor = extractors.read_spikeglx(os.path.split(file_path)[0], "imec.ap")
+        elif ephys_recording_record["acq_software"] == "OpenEphys":
+            extractor = extractors.read_openephys(file_path, stream_id="0")
+        else:
+            raise ValueError(
+                f"unsupported acq_software type: {ephys_recording_record['acq_software']}"
+            )
+
+        conversion_kwargs = gains_helper(extractor.get_channel_gains())
+
+        if end_frame is not None:
+            extractor = extractor.frame_slice(0, end_frame)
+
+        recording_channels_by_id = (
+            probe.ElectrodeConfig.Electrode() & ephys_recording_record
+        ).fetch("electrode")
+
+        nwbfile.add_acquisition(
+            pynwb.ecephys.ElectricalSeries(
+                name=f"ElectricalSeries{ephys_recording_record['insertion_number']}",
+                description=str(ephys_recording_record),
+                data=SpikeInterfaceRecordingDataChunkIterator(extractor),
+                rate=ephys_recording_record["sampling_rate"],
+                starting_time=(
+                    ephys_recording_record["recording_datetime"]
```

Comment:
I see a `recording_datetime` value in the Open Ephys reader, but I don't see it in the `ephys.EphysRecording` table where `ephys_recording_record` comes from. On run, I get a KeyError.

Reply:
Thanks @CBroz1.  I will submit a pull request shortly adding `recording_datetime` and `recording_duration` to the `ephys_acute` and `ephys_chronic` modules.  I believe that is the primary difference.  As pull request #54 is not yet merged, I would suggest removing `ephys_no_curation` from this pull request.  I have tested locally and the `ephys_acute` module works with this NWB export function and pytests.

Reply:
Just a note.  This export function will not work for the `ephys_acute` or `ephys_chronic` modules as the export function was designed for the `ephys_no_curation` module which does not have a `curation_id`.